firedrake/adjoint/ensemble_reduced_functional.py

```diff
@@ -53,15 +61,72 @@ class EnsembleReducedFunctional(ReducedFunctional):
     works, please refer to the `Firedrake manual
     <https://www.firedrakeproject.org/parallelism.html#id8>`_.
     """
-    def __init__(self, J, control, ensemble, scatter_control=True):
-        super(EnsembleReducedFunctional, self).__init__(J, control)
+    def __init__(self, J, control, ensemble, scatter_control=True,
+                 gather_functional=None):
         self.ensemble = ensemble
+        if isinstance(J, list):
+            assert isinstance(control, list), "Controls should also be a list."
+            assert len(control) % len(J) == 0, "Controls and J have mismatching lengths."
+            self.controls_per_J = len(control) // len(J)
+            self.functional = J
+            self.controls = Enlist(control)
+            self.Jhats = []
+            self.sizes = self.ensemble.ensemble_comm.allgather(len(J))
+            for i in range(len(J)):
+                i0 = self.controls_per_J*i
+                i1 = self.controls_per_J*(i+1)
+                self.Jhats.append(ReducedFunctional(J[i], control[i0:i1]))
```

Comment:
Perhaps would be better to set as a list of tuples, where each tuple corresponds to a set of controls for each `J` element. 

Then, you can iterate through `J` and `control` simultaneously:
```
for i in range(len(J)):
    self.Jhats.append(ReducedFunctional(J[i], control[i]))

```